### PR TITLE
Modify installation script to be idempotent

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

I discovered issues in my dev environment were due to needed tools not being installed.

The `dev.sh` script was short circuiting installation by assuming everything was present if `task` was present, resulting in the silent omission of new tooling.

I've changed the `install-dependencies` script to be idempotent, so it can be routinely run in full and it won't make any changes unless needed. It also now has `--verbose` (`-v`) and `--force` (`-f`) options so that reinstallation can be forced if desired.

**How does this PR make you feel**:
gif

### CODE REVIEW

File: .devcontainer/install-dependencies.sh [hunk 1]
```diff
@@ -49,62 +111,95 @@ export GOPATH=$TMPDIR
...
+        curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b "$TOOL_DEST" v1.45.2 2>&1
+    fi
 fi
 
-# Install go-task (task runner)
-echo "Installing go-task…"
-curl -sL "https://github.com/go-task/task/releases/download/v3.7.0/task_linux_amd64.tar.gz" | tar xz -C "$TOOL_DEST" task
+# Install Task
+write-verbose "Checking for go-task"
+if [ "$FORCE" == true ] || [ ! -f "$TOOL_DEST/task" ]; then
```
Comment (Porges) [inline .devcontainer/install-dependencies.sh:155]: a helper like `should_install "$TOOL_DEST/xyz"` might be helpful

File: .devcontainer/install-dependencies.sh [hunk 2]
```diff
@@ -24,17 +75,28 @@ else
...
+# Ensure we have the right version of GO
+
 if ! command -v go > /dev/null 2>&1; then
-    echo "Go must be installed manually: https://golang.org/doc/install"
+    write-error "Go must be installed manually; see https://golang.org/doc/install"
+    exit 1
+fi
+
+GOVERREQUIRED="go1.18"
+GOVERACTUAL=`go version | { read _ _ ver _; echo $ver; }`
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply .devcontainer/install-dependencies.sh]: Fixed.

File: .devcontainer/install-dependencies.sh (hunk 1 above)
Comment (theunrepentantgeek) [reply .devcontainer/install-dependencies.sh]: Good idea - makes things more declarative in a good way.

**Review by matthchr — CHANGES_REQUESTED** [review]

File: .devcontainer/install-dependencies.sh [hunk 3]
```diff
@@ -49,62 +111,108 @@ export GOPATH=$TMPDIR
...
+    fi
+
+    # No, installation is not
+    return 1
+}
+
+
+# go-install() is a helper function to trigger `go install` 
+# if a command is missing OR if we're using --force
+go-install() {
```
[the remark this answers is not in the document]
Comment (matthchr) [reply .devcontainer/install-dependencies.sh:133]: Ah, some of the go installs are pretty slow. See my comment below about reversing the default of the script

[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply .devcontainer/install-dependencies.sh:133]: The `go install` command will run every time; some of them can take a few minutes to download/compile.

File: .devcontainer/install-dependencies.sh [hunk 4]
```diff
@@ -49,62 +111,108 @@ export GOPATH=$TMPDIR
...
+        curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b "$TOOL_DEST" v1.45.2 2>&1
+    fi
 fi
 
-# Install go-task (task runner)
-echo "Installing go-task…"
-curl -sL "https://github.com/go-task/task/releases/download/v3.7.0/task_linux_amd64.tar.gz" | tar xz -C "$TOOL_DEST" task
+# Install Task
+write-verbose "Checking for go-task"
+if should-install "$TOOL_DEST/task"; then
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply .devcontainer/install-dependencies.sh:169]: As discussed offline, I've changed the flag to `--skip-installed` so the default behavior (for people who don't use a flag) is to install.

**Review by matthchr — APPROVED** [review]

**Pull request merged**